cli/azd/pkg/prompt/prompter.go

```diff
@@ -111,6 +112,51 @@ func (p *DefaultPrompter) PromptLocation(
 	return loc, nil
 }
 
+// PromptResource uses the console (or external) prompter to allow the user to select a resource
+// of optional type `resourceType`. The selected resource Name will be returned.
+// If no resources of that type were found, an empty string is returned.
+//
+// The Name can be used with the ARM or Bicep template function `reference` or in an existing resource template's name
+// to get provisioned state data from a resource, or passed to the `resourceId` function to get the full resource ID
+// if you know the `resourceType`.
+func (p *DefaultPrompter) PromptResource(
+	ctx context.Context,
+	msg string,
+	resourceType string,
+) (string, error) {
+	options := azapi.ListResourcesOptions{
+		ResourceType: resourceType,
+	}
+	resources, err := p.resourceService.ListResources(ctx, p.env.GetSubscriptionId(), &options)
+	if err != nil {
+		return "", fmt.Errorf("listing resources: %w", err)
+	}
+	if len(resources) == 0 {
+		return "", nil
+	}
```

Comment:
I think this should return an error. Since the intention of the method is to prompt, the fact of an empty list is a blocker to do the prompting, so it should return an error to the caller and report that there's no resources to prompt for.

Reply:
FWIW, I ended up leveraging what @heaths had here in #4943. 

While engaging more deeply with the code, I appreciated @heaths implementation here, since whether the resource type is optional is left up to the caller to decide. It allows for a shorter and user-friendly error message.

Reply:
FWIW, my main thought with separating optional out was to also allow more permutations for other selector types. I've seen some APIs that have things like `foo`, `fooOptional`, `bar`, `barOptional`, etc., and the maintenance of and UX for all those is cumbersome. I like to separate "dials".